control/controller.go

```diff
@@ -377,32 +323,37 @@ func (q *Query) Cancel() {
 // The query may also have an error during execution so the Err()
 // function should be used to check if an error happened.
 func (q *Query) Results() <-chan flux.Result {
-	return q.ready
+	return q.results
 }
 
 // Done signals to the Controller that this query is no longer
 // being used and resources related to the query may be freed.
 //
+// The Results method must have been completely read before calling
+// this method either by the query executing, being canceled, or
+// an error occurring.
```

Comment:
I think you wanted to remove the comment below.  

Anyhow, why should the result be consumed before calling Done()?
Is it because `pump` would block?

Couldn't `Done` call `cancel` on `currentCtx`, so that pump stops consuming `exec.Results()`. Maybe, `pump` would also block on sending the result on `q.results`...

Could you provide a `ConsumeResults(q Query, fn func(flux.Result) error) error` as a helper to summarize the logic of consuming a result by applying a function and check for `q.Err()` and calling `q.Done()`?

Reply:
I think I messed this up. I'll make it so done will cancel the pumping of results.